File: Python3/104. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = []
        depth = 0
        
        if not root:
            return depth
        
        # the number indicates its depth
        stack.append((1, root))
        
        while stack:
            cur_depth, root = stack.pop()
            if root:
                depth = max(depth, cur_depth)
                stack.append((cur_depth+1, root.left))
                stack.append((cur_depth+1, root.right))
                
        return depth
        
                

File: Java/104. Maximum Depth of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        int depth = 0;
        Stack<TreeNode> stack = new Stack<>();
        Stack<Integer> level = new Stack<>();
        
        if(root == null){
            return depth;
        }
        
        stack.push(root);
        level.push(1);
        
        while(!stack.isEmpty()){
            root = stack.pop();
            int d = level.pop();
            
            if(root != null){
                depth = Math.max(depth, d);
                stack.push(root.left);
                level.push(d+1);
                stack.push(root.right);
                level.push(d+1);
            }
        }
        
        return depth;
        
    }
}